Source/Core/InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServerProto.h

```diff
@@ -0,0 +1,262 @@
+  MessageHeader header;
+  u32 message_type;
+};
+
+struct VersionResponse
+{
+  static constexpr auto FROM = SERVER;
+  static constexpr auto TYPE = 0x100000U;
+  MessageHeader header;
+  u32 message_type;
+  u16 max_protocol_version;
+  u8 padding[2];
+};
+
+struct ListPorts
+{
+  static constexpr auto FROM = CLIENT;
+  static constexpr auto TYPE = 0x100001U;
+  MessageHeader header;
+  u32 message_type;
+  u32 pad_request_count;
+  u8 pad_id[4];
+};
+
+struct PortInfo
+{
+  static constexpr auto FROM = SERVER;
+  static constexpr auto TYPE = 0x100001U;
+  MessageHeader header;
+  u32 message_type;
+  u8 pad_id;
+  DsState pad_state;
+  DsModel model;
+  DsConnection connection_type;
+  u8 pad_mac_address[6];
+  DsBattery battery_status;
+  u8 padding;
+};
+
+struct PadDataRequest
+{
+  static constexpr auto FROM = CLIENT;
+  static constexpr auto TYPE = 0x100002U;
+  MessageHeader header;
+  u32 message_type;
+  RegisterFlags register_flags;
+  u8 pad_id_to_register;
+  u8 mac_address_to_register[6];
+};
+
+struct PadDataResponse
+{
+  static constexpr auto FROM = SERVER;
+  static constexpr auto TYPE = 0x100002U;
+  MessageHeader header;
+  u32 message_type;
+  u8 pad_id;
+  DsState pad_state;
+  DsModel model;
+  DsConnection connection_type;
+  u8 pad_mac_address[6];
+  DsBattery battery_status;
+  u8 active;
+  u32 hid_packet_counter;
+  u8 button_states1;
+  u8 button_states2;
+  u8 button_ps;
+  u8 button_touch;
+  u8 left_stick_x;
+  u8 left_stick_y_inverted;
+  u8 right_stick_x;
+  u8 right_stick_y_inverted;
+  u8 button_dpad_left_analog;
+  u8 button_dpad_down_analog;
+  u8 button_dpad_right_analog;
+  u8 button_dpad_up_analog;
+  u8 button_square_analog;
+  u8 button_cross_analog;
+  u8 button_circle_analog;
+  u8 button_triangle_analog;
+  u8 button_r1_analog;
+  u8 button_l1_analog;
+  u8 trigger_r2;
+  u8 trigger_l2;
+  Touch touch1;
+  Touch touch2;
+  u64 timestamp_us;
+  float accelerometer_x_g;
+  float accelerometer_y_g;
+  float accelerometer_z_inverted_g;
+  float gyro_pitch_deg_s;
+  float gyro_yaw_deg_s;
+  float gyro_roll_deg_s;
+};
+
+struct FromServer
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionResponse version_response;
+    MessageType::PortInfo port_info;
+    MessageType::PadDataResponse pad_data_response;
+  };
+};
+
+struct FromClient
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionRequest version_request;
+    MessageType::ListPorts list_ports;
+    MessageType::PadDataRequest pad_data_request;
+  };
+};
+}  // namespace MessageType
+
+static inline u32 CRC32(const void* buffer, unsigned length)
+{
+  return crc32(crc32(0L, Z_NULL, 0), (const Bytef*)buffer, length);
+}
+
+template <typename MsgType>
+struct Message : public MsgType
+{
+  Message() {}
+
+  Message(u32 source_uid)
+  {
+    memset(this, 0, sizeof *this);
```

Comment:
You can simply make the data member part of the class itself instead of using inheritance, and give it a default member initializer instead of needing a memset call, as far as I can tell. This way, the class also doesn't invoke undefined behavior if changes to the structs ever make them non-trivially copyable.

Reply:
I changed `memset(this, 0, sizeof *this);` into a base class initializer: `Message(u32 source_uid) : MsgType{} { ... }`. This should be safer and do the same thing, shoudn't it?

After trying out using a data member in stead of inheritance, I decided to stick with inheritance. The main reason is using the class becomes a little messy. Example: (namespaces abbreviated for clarity)

With inheritance:
```
Message<MessageType::PadDataRequest> data_req(m_client_uid);
data_req.register_flags = RegisterFlags::PadID;
data_req.pad_id_to_register = m_index;
data_req.Finish();
```
vs with data member:
```
Message<MessageType::PadDataRequest> data_req(m_client_uid);
data_req.m_message.register_flags = RegisterFlags::PadID;
data_req.m_message.pad_id_to_register = m_index;
data_req.Finish();
```
You have two objects (`data_req` and `data_req.m_message`) to use in stead of one. When subsequently calling `m_socket.send()` there is confusion of which object to pass, I think.

Reply:
I would disagree with leaving it as is. Especially when it makes it more difficult to unintentionally introduce undefined behavior within the type system and keeps the class itself flexible in terms of layout requirements.

If there's a potential spot for confusion, then this is either resolved by introducing a function that always does the right thing, and/or introducing a comment that clarifies that. Conflating the representation of a message with its raw data is not a very good idea. memsetting through the this pointer is generally advised against, and for good reason.

Reply:
As you can see at the top of my comment, I replaced memset(this ...) with a base class initializer. Shouldn't that solve this sticky point?

Reply:
Partially, but not entirely, because the issue of conflating the message type with its contained data is still a thing. I would still very much rather have the data within the message be separated from the actual type itself in a non-inherited context.

Reply:
> Conflating the representation of a message with its raw data is not a very good idea.

I'm having trouble understanding this problem. Can you give an example of a _practical_ worst case scenario with this specific implementation? I just don't like principles getting in the way of practical engineering without a good reason.

Reply:
The entire purpose is to keep the representation of the data separate from the entity that actually contains it (e.g. consider: `std::array`, `std::chrono::duration`, `std::vector`, [insert any other kind of class that contains data]). This way, the interface type is allowed to change without potentially/accidentally breaking the contained data set (especially when using the container as if it were the raw underlying data set).

If you're truly that uncomfortable with the notion of that, I can make the change myself after this is merged.

Reply:
I have changed inheritance to data member.